cmd/admin-revoker/main.go

```diff
@@ -211,7 +237,117 @@ func (r *revoker) revokeByReg(ctx context.Context, regID int64, reasonCode revoc
 }
 
 func (r *revoker) revokeMalformedBySerial(ctx context.Context, serial string, reasonCode revocation.Reason) error {
-	return r.revokeCertificate(ctx, core.Certificate{Serial: serial}, reasonCode)
+	return r.revokeCertificate(ctx, core.Certificate{Serial: serial}, reasonCode, false)
+}
+
+// blockByPrivateKey blocks future issuance for certificates with a a public key
+// matching the SubjectPublicKeyInfo hash generated from the PublicKey embedded
+// in privateKey. The embedded PublicKey will be verfied as an actual match for
+// the provided private key before any blocking takes place. This method does
+// not revoke any certificates directly. However, 'bad-key-revoker', which
+// references the 'blockedKeys' table, will eventually revoke certificates with
+// a matching SPKI hash.
+func (r *revoker) blockByPrivateKey(ctx context.Context, privateKey crypto.Signer) error {
+	err := privatekey.Verify(privateKey)
+	if err != nil {
+		return err
+	}
+
+	spkiHash, err := getPublicKeySPKIHash(privateKey.Public())
+	if err != nil {
+		return err
+	}
+
+	u, err := user.Current()
+	if err != nil {
+		return err
+	}
+
+	req := &sapb.AddBlockedKeyRequest{
+		KeyHash:   spkiHash,
+		Added:     r.clk.Now().UnixNano(),
+		Source:    "admin-revoker",
+		Comment:   fmt.Sprintf("blocked by %s", u),
+		RevokedBy: 0,
+	}
+
+	_, err = r.sac.AddBlockedKey(ctx, req)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// revokeByPrivateKey revokes all certificates with a public key matching the
+// SubjectPublicKeyInfo hash generated from the PublicKey embedded in
+// privateKey. The embedded PublicKey will be verfied as an actual match for the
+// provided private key before any revocation takes place. The provided key will
+// not be added to the 'blockedKeys' table, this is done to avoid a race between
+// 'admin-revoker' and 'bad-key-revoker'. You MUST call blockByPrivateKey after
+// calling this function, on pain of violating the BRs.
+func (r *revoker) revokeByPrivateKey(ctx context.Context, privateKey crypto.Signer) error {
+	err := privatekey.Verify(privateKey)
+	if err != nil {
+		return err
+	}
+
+	spkiHash, err := getPublicKeySPKIHash(privateKey.Public())
+	if err != nil {
+		return err
+	}
+
+	matches, err := r.getCertsMatchingSPKIHash(spkiHash)
+	if err != nil {
+		return err
+	}
+
+	for i, match := range matches {
+		err := r.revokeBySerial(ctx, match, revocation.Reason(1), true)
+		if err != nil {
+			return fmt.Errorf(
```

Comment:
Here we need to deal with the possibility that a certificate has already been revoked (for instance, if we ran this once before and interrupted it).

In the sa, if that happens we return an InternalServerError: https://github.com/letsencrypt/boulder/blob/06ba17c2be6ffffd5cd2c0ffd5f79e066897736f/sa/sa.go#L1714-L1741. That's because we expect RA to already have called GetCertificateStatus and checked that the certificate is not revoked (there's a slight [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issue there, in that it could be revoked between GetCertificateStatus and RevokeCertificate).

There's two ways to go about fixing this:

 - Call GetCertificateStatus before revoking in this code path, too.
 - Add a new berrors code AlreadyRevoked, to mirror probs.AlreadyRevoked. Modify SA to detect the difference between "doesn't exist" and "already revoked" (not sure exactly how to do this nicely), and return AlreadyRevoked in that case.

I think the first one is much simpler and recommend it.

Reply:
Thanks for pointing this out. I went ahead and implemented option 1.